### PR TITLE
add end-to-end test for emulator function triggers

### PR DESCRIPTION
**Pull request by IanWyszynski**
### Description

This tests checks that RTDB and Firestore emulator write triggers work, and that the admin SDK can be used from within such a triggered function to write back to either the RTDB emulator or the Firestore emulator. See:

- https://github.com/firebase/firebase-tools/pull/1347
- https://github.com/firebase/firebase-tools/pull/1411

for more context.

I created a new firebase project called `fir-tools-testing` to run this test, but I'm not sure if that's necessary (or if we have a different project we use for CI tests -- if that is the case I will migrate this test over). 

This test is just a starting point, I plan to add more coverage in the future by adding new driver programs in the `firebase-tools/triggers-end-to-end-tests` directory in the future. 

TODO: I have yet to create/test a new travis job for this test.
TODO: This PR depends on changes made in https://github.com/firebase/firebase-tools/pull/1411, so I will rebase this onto master once https://github.com/firebase/firebase-tools/pull/1411 goes in.
	 
### Sample Commands

`cd firebase-tools/scripts/triggers-end-to-end-tests && ./test.js`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -82,6 +82,7 @@
     "fs-extra": "^0.23.1",
     "glob": "^7.1.2",
     "google-auto-auth": "^0.7.2",
+    "grpc": "^1.21.1",
```
Comment (samtstern) [inline package.json]: does `firebase-tools` itself need this dependency or should this be part of the `package.json` of the individual integration tests?

File: scripts/triggers-end-to-end-tests/functions/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,123 @@
+{
```
Comment (samtstern) [inline scripts/triggers-end-to-end-tests/functions/.eslintrc.json]: Let's not add new linting for the end-to-end test source.  We can either ignore it or extend the root `lint` task to look at these files.

File: scripts/triggers-end-to-end-tests/functions/package-lock.json [hunk 3]
```diff
@@ -0,0 +1,1927 @@
+{
```
Comment (samtstern) [inline scripts/triggers-end-to-end-tests/functions/package-lock.json]: We might want to not check in `package-lock.json` files in these end to end tests.  Seems like churn for not much benefit (whereas the root one is more useful)

File: package-lock.json [hunk 4]
```diff
@@ -230,7 +230,6 @@
           "resolved": "https://registry.npmjs.org/gaxios/-/gaxios-1.8.4.tgz",
           "integrity": "sha512-BoENMnu1Gav18HcpV9IleMPZ9exM+AvUjrAOV4Mzs/vfz2Lu/ABv451iEXByKiMPn2M140uul1txXCg83sAENw==",
           "dev": true,
-          "optional": true,
```
Comment (ryanpbrewster) [inline package-lock.json]: these changes seem unrelated, are you on a different version of `npm`?

File: scripts/triggers-end-to-end-tests/database.rules.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "rules": {
+    ".read": false,
+    ".write": false
```
Comment (ryanpbrewster) [inline scripts/triggers-end-to-end-tests/database.rules.json]: do we need rules at all? the emulator starts with open rules

File: scripts/triggers-end-to-end-tests/firebase.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{
+  "database": {
+    "rules": "database.rules.json"
+  },
+  "firestore": {
+    "rules": "firestore.rules",
+    "indexes": "firestore.indexes.json"
+  },
```
Comment (ryanpbrewster) [inline scripts/triggers-end-to-end-tests/firebase.json]: I assume we don't use firestore

**Comment by ryanpbrewster** [comment]
and how to run it

File: scripts/triggers-end-to-end-tests/firebase.json (hunk 6 above)
Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/firebase.json]: We are not, but this blob seems necessary for creating the firestore emulator client. Judging by the error I see when I remove this, creating a Firestore client taps into some pool of cached clients that (I assume) sources this config somewhere.

File: .travis.yml [hunk 7]
```diff
@@ -12,6 +12,13 @@ jobs:
       before_script: ./scripts/decrypt-app-credentials.sh
       script: ./scripts/test-hosting.sh
       after_script: skip
+  include:
+    - stage: emulator triggers end to end test
+      if: repo == head_repo OR type == push
+      node_js: "8"
+      before_script: skip
+      script: ./scripts/triggers-end-to-end-tests/test.js
```
Comment (samtstern) [inline .travis.yml]: Right now fails with: 
```
$ ./scripts/triggers-end-to-end-tests/test.js
unable to read firebase.json Error: ENOENT: no such file or directory, open 'firebase.json'
```

I think we want to add a `./end-to-end.sh` and then have that look something like:
```
cd ./scripts/triggers-end-to-end-tests
node test.js
```

That will also allow us to add more tests over time

File: scripts/triggers-end-to-end-tests/firebase.json (hunk 6 above)
Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/firebase.json]: This config is also used to populate the firebase-tools `EmulatorRegistry`, and the functions emulator thinks the Firestore emulator is not running if this entry doesn't exist.

Comment (samtstern) [reply scripts/triggers-end-to-end-tests/firebase.json]: Yeah if you want `emulators:start` to start firestore you need a `firestore` block in your `firebase.json`. 

We could definitely remove the `indexes` sub-entry and the `indexes.json` file though.

File: scripts/triggers-end-to-end-tests/firebase.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  "functions": {
+    "predeploy": [
```
Comment (samtstern) [inline scripts/triggers-end-to-end-tests/firebase.json]: We should drop this whole `predeploy` block since we don't deploy.

File: scripts/triggers-end-to-end-tests/functions/index.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+});
+
+exports.firestoreReaction = functions.firestore.document(START_DOCUMENT_NAME).onWrite(async (change, ctx) => {
+    console.log(FIRESTORE_FUNCTION_LOG);
+    /*
+     * Write back a completion timestamp to the firestore emulator. The test
+     * driver program checks for this by querying the firestore emulator
+     * directly.
+     */
+    const ref = admin.firestore().doc(END_DOCUMENT_NAME + "_from_firestore");
```
Comment (samtstern) [inline scripts/triggers-end-to-end-tests/functions/index.js]: What's the purpose of the `_from_firestore` and `_from_database` suffixes?

Comment (samtstern) [reply scripts/triggers-end-to-end-tests/functions/index.js]: Ohhh I see we're writing to RTDB from Firestore functions (and vice-versa).

Does that test any particular concern?

File: scripts/triggers-end-to-end-tests/test.js [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    try {
+      config = JSON.parse(data);
+    } catch (err) {
+      console.log("malformed firebase.json " + err);
+    }
+    done(null, config);
+  });
+}
+
+function runTest(config)
```
Comment (samtstern) [inline scripts/triggers-end-to-end-tests/test.js]: Could we use a testing library (mocha, ideally) to run this?  A few benefits:
1. Could use real `assert` statements below instead of writing your own checks.
2. Support for cleaner start/cleanup with `before()` and `after()`
3. No need for `sleep()` since they support async tests with timeouts.

If you haven't used mocha before and would find this daunting let me know, I am happy to make that change as a follow-up.

File: scripts/triggers-end-to-end-tests/database.rules.json (hunk 5 above)
Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/database.rules.json]: It looks like we don't need either `firestore.rules` or `database.rules.json`, but `firestore.rules` needs to be present in `firebase.json` even if the actual file doesn't exist.

File: scripts/triggers-end-to-end-tests/functions/index.js (hunk 9 above)
Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/functions/index.js]: The suffixes indicate whether the object was written from a firestore or realtime database trigger.

Comment (samtstern) [reply scripts/triggers-end-to-end-tests/functions/index.js]: Is there any reason to suspect that writing to RTDB from a Firestore function would be different than writing to RTDB from a RTDB function?

IMO it would be sufficient just to test RTDB<-->Functions<-->RTDB and Firestore<-->Functions<-->Firestore

Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/functions/index.js]: Nothing specific, but the key change we missed in the original PR was having the functions emulator configure the "in-cloud-function" admin sdk to point at the database emulator. I thought ensuring this worked for both emulators was a valuable exercise. I think covering this specific use-case in our tests might also be valuable for developers looking to test a firestore -> database or database -> firestore migration/sync pipelines.

Comment (samtstern) [reply scripts/triggers-end-to-end-tests/functions/index.js]: SGTM since you already wrote it and I can see no harm in this.

File: scripts/triggers-end-to-end-tests/test.js (hunk 10 above)
Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/test.js]: Done, only we still do need sleeps in places where we need to wait for parallel emulator subprocesses to start up.

File: scripts/triggers-end-to-end-tests/run.spec.js [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+      if (err) {
+        done(new Error("error reading firebase.json: " + err));
+        return;
+      }
+      test = new TriggerEndToEndTest(config);
+      done();
+    });
+  });
+
+  before(function (done) {
```
Comment (samtstern) [inline scripts/triggers-end-to-end-tests/run.spec.js]: I would recommend pushing these two `before` blocks into one since I don't think there is a guarantee on calling order (besides that it's just convention to have a single before and a single after)

Comment (samtstern) [reply scripts/triggers-end-to-end-tests/run.spec.js]: Oh now that I am reading I see that there are four `before()` blocks ... can you explain what you were trying to do?

Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/run.spec.js]: I think the blocks are called in the order they were defined: https://mochajs.org/#run-cycle-overview. I was using this assumption to isolate different steps of the setup process.

Comment (IanWyszynski) [reply scripts/triggers-end-to-end-tests/run.spec.js]: I have merged the setup code into a single before(...) and chained the individual setup steps together with an `async.series` call.

File: .travis.yml [hunk 12]
```diff
@@ -12,6 +12,13 @@ jobs:
       before_script: ./scripts/decrypt-app-credentials.sh
       script: ./scripts/test-hosting.sh
       after_script: skip
+  include:
```
Comment (samtstern) [inline .travis.yml]: I don't see the "hosting functional test" running on this PR .... I wonder if your `include` is wiping out the previous value and instead you just want to add another `-stage` to the hosting test.

**Comment by samtstern** [comment]
@IanWyszynski #1411 has been merged into master.
